Feature/add ukhpi pricefeed

krisoshea-eth opened the pull request:
feat(pricefeed): adds UKHPI pricefeed to bot config

 
**Motivation**

This PR adds the UKHPI pricefeed start-up script to the uma-protocol bot config start-up process.

**Summary**

In order to achieve the above, an additional 'else-if' block was added to the CreatePriceFeed.ts file and the  HmLandRegistryPriceFeed.ts file was created in order to enable the bots to query the Hm Land Registry API for UKHPI price data.


**Testing**

- [ X]  Ran end-to-end test, running the code as in production

krisoshea-eth committed:
Feature/Add-UKHPI-pricefeed

This commit adds the HmLandRegistryPriceFeed.ts file to the forked UMA protocol repo in order to support the creation of UKHPI price feeds for the bots. This commit also adds an additional 'else-if' block to the CreatePriceFeed.ts file in order to reflect these changes.

krisoshea-eth committed:
Update HmLandRegistryPriceFeed.ts

This commit updates the HmLandRegistryPriceFeed.ts file

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+  private lastUpdateTime: number | null = null;
+
+  /**
+   * @notice Constructs the TwelveDataApiPriceFeed.
```
can we change this comment?

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+  /**
+   * @notice Constructs the TwelveDataApiPriceFeed.
+   * @param {Object} logger Winston module used to send logs.
+   * @param {String} index String used in query to fetch index data, i.e. "URTH"
```
update comment for UKHPI

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+        });
+
+     // checks if the UKHPI value is being returned correctly   
+      newHistoricalPricePeriods.forEach((item: any) => {
```
can remove this loop

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+  constructor(
+    private readonly logger: Logger,
+    private readonly index: string,
+    private readonly lookback: number, // lookback should ideally be 90 days
```
let's update this comment to 120 days, since ideal lookback needs to be atleast 120 days for UKHPI. Using a lookback of 90 days, gives no results.

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+
+  FILTER (?region = <http://landregistry.data.gov.uk/id/region/united-kingdom>)
+}
+ORDER BY DESC(?date)
```
Change this to `ORDER BY ASC(?date)`. Code is returning incorrect current price and values if we use DESC.

The below matching code works with ASC order data only. See this comment _// historicalPricePeriods are ordered from oldest to newest._

KohliSuraj commented on packages/financial-templates-lib/src/price-feed/HmLandRegistryPriceFeed.ts:
```diff
@@ -0,0 +1,253 @@
+    // 5. Store results.
+    this.currentPrice = newHistoricalPricePeriods[newHistoricalPricePeriods.length - 1].value;
+    this.priceHistory = newHistoricalPricePeriods;
+    this.lastUpdateTime = currentTime;
```
I tested the price feed be adding these lines here, and giving different lookbacks like 120 days, 365 days etc.

```
    // TESTING Update price function
    console.log("DEBUG getCurrentPrice: ", this.getCurrentPrice()?.toString());
    console.log("DEBUG getLastUpdateTime: ", this.getLastUpdateTime());
    console.log("DEBUG getHistoricalPrice: 1st May", (await this.getHistoricalPrice(1682928606))?.toString());
    console.log("DEBUG getHistoricalPrice: 5th March", (await this.getHistoricalPrice(1678003806))?.toString());
    console.log("DEBUG getHistoricalPrice: 9th Feb", (await this.getHistoricalPrice(1675930206))?.toString());
    console.log("DEBUG getHistoricalPrice: 1st Jan", (await this.getHistoricalPrice(1672560606))?.toString());
    console.log("DEBUG getHistoricalPrice: 28th Dec", (await this.getHistoricalPrice(1672215006))?.toString());
```

krisoshea-eth replied:
I repeated this exact testing process locally after making the above changes you recommended and everything appears to be working perfectly now. I think it is good to merge :)

KohliSuraj replied:
Perfect 💯

KohliSuraj reviewed (changes requested):
@krisoshea-eth Excellent attempt. Just 1 small bug!

krisoshea-eth committed:
Update HmLandRegistryPriceFeed.ts file to include recommended PR changes

This commit updates the HmLandRegistryPriceFeed.ts file based on the advice and feedback provided by Suraj on the recent UKHPI pricefeed implementation PR
